File: ios/HeightMeasure/Measurement/MeasurementSession.swift
import ARKit
import Combine

enum TapState {
    case idle
    case bottomPlaced(point: SIMD3<Float>)
    case complete(result: MeasurementResult)
}

class MeasurementSession: ObservableObject {
    @Published var tapState: TapState = .idle
    @Published var completedResult: MeasurementResult?

    private let categories: [HeightCategory]

    init(categories: [HeightCategory] = HeightCategory.defaults) {
        self.categories = categories
    }

    func recordTap(worldPoint: SIMD3<Float>) {
        switch tapState {
        case .idle:
            tapState = .bottomPlaced(point: worldPoint)

        case .bottomPlaced(let bottom):
            let vertical = HeightCalculator.verticalHeight(bottom: bottom, top: worldPoint)
            let euclidean = HeightCalculator.euclideanDistance(a: bottom, b: worldPoint)
            let result = MeasurementResult(
                id: UUID(),
                verticalHeight: vertical,
                euclideanDistance: euclidean,
                category: HeightCategory.category(for: vertical, categories: categories),
                bottomPoint: bottom,
                topPoint: worldPoint,
                timestamp: Date()
            )
            tapState = .complete(result: result)
            completedResult = result

        case .complete:
            break
        }
    }

    func reset() {
        tapState = .idle
        completedResult = nil
    }
}

File: unity/HeightMeasure/Assets/Scripts/Measurement/MeasurementSession.cs
using UnityEngine;
using System;

public class MeasurementSession : MonoBehaviour
{
    public enum State { Idle, BottomPlaced, Complete }

    public State CurrentState { get; private set; } = State.Idle;
    public Vector3 BottomPoint { get; private set; }
    public MeasurementResult LastResult { get; private set; }

    public event Action<State> OnStateChanged;
    public event Action<MeasurementResult> OnMeasurementComplete;

    public void RecordTap(Vector3 worldPoint)
    {
        switch (CurrentState)
        {
            case State.Idle:
                BottomPoint = worldPoint;
                Transition(State.BottomPlaced);
                break;

            case State.BottomPlaced:
                float vertical  = HeightCalculator.VerticalHeight(BottomPoint, worldPoint);
                float euclidean = HeightCalculator.EuclideanDistance(BottomPoint, worldPoint);
                LastResult = new MeasurementResult
                {
                    verticalHeight   = vertical,
                    euclideanDistance = euclidean,
                    category         = HeightCategory.Get(vertical),
                    bottomPoint      = BottomPoint,
                    topPoint         = worldPoint,
                    timestamp        = DateTime.Now,
                };
                Transition(State.Complete);
                OnMeasurementComplete?.Invoke(LastResult);
                break;

            case State.Complete:
                break;
        }
    }

    public void Reset()
    {
        LastResult = null;
        Transition(State.Idle);
    }

    void Transition(State next)
    {
        CurrentState = next;
        OnStateChanged?.Invoke(next);
    }
}
